[SYCL][Doc] Add InvokeSIMD and Uniform extensions (#3445)
intel/llvm: sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc

converting arguments and return values between scalar and SIMD types as
appropriate.

[source, c++]
----
namespace sycl {
namespace ext {
namespace oneapi {

template <typename Callable, typename...T>
auto invoke_simd(sub_group sg, Callable&& f, T... args);

} // namespace oneapi
} // namespace ext
} // namespace sycl
----

The `Callable` passed to `invoke_simd` must be a C++ callable that can be
invoked with the same number of arguments specified in the `args` parameter
pack. `Callable` may be a function object, a lambda,
or a function pointer (if the device supports `SPV_INTEL_function_pointers`).
`Callable` must be an immutable callable with the same type and state for all
work-items in the sub-group, otherwise behavior is undefined.

Each argument to the invoked `Callable` must be a scalar type, a SIMD type,
or a `std::tuple` of these types. The length of each SIMD type must match the
sub-group size of the kernel; calling a SIMD function with mismatched lengths
is an error, and an implementation must issue a diagnostic.

To enable overloading in cases where the `Callable` accepts either a scalar
type or a SIMD type, we leverage the `sycl::ext::oneapi::uniform` wrapper

> type from the `SYCL_EXT_ONEAPI_UNIFORM` extension.

Reviewer: Even we discussed this in email, I would like to state my opinion here as I'm still not convinced by our discussion.  I believe having this special sycl::ext::oneapi::uniform is not needed in general. The compiler can match several possible callees by matching their data type, taking the "underlying" data type from vectors, so f(int ) and f(simd<int>) will both match. And then generate the call to the variant selected by uniformity analysis.  That strategy is perfectly aligned with our function pointers.

Author: We don't want the compiler to do the matching, and we don't want the compiler to do any uniformity analysis.  The purpose of `uniform` is to give developers complete control over matching (using regular C++ overload rules), and to let developers assert uniformity in situations where the compiler can't possibly prove it.

Reviewer: The first statement is a little unclear. Who then will do matching?.. The suggested way just simplifies that matching.  
I'm not against this simplification, but I really don't like a "fully autonomous" uniform specification that can appear anywhere in the program.  That can lead to very hard-detectible mistakes, when user thinks the value is uniform but really it's not. E.g. a function parameter. When a user creates a function and uses it himself it may be always uniform.  But that can change when some one else uses that function.  I'll leave a comment in the uniform specification.

Author: Everything is handled by the C++ overloading mechanism and the type system.  A developer cannot accidentally declare something to be `uniform` by calling a function -- they have to call the `uniform` constructor explicitly.

Here's an example that might help:

```c++
// Function with varying (T) and uniform (uniform<T>) overloads
void foo(int x);
void foo(uniform<int> x);

// Function with varying (T) argument only
void bar(int x);

// Function with uniform (uniform<T>) argument only
void baz(uniform<T> x);

int main()
{
    int x;                // x is assumed varying
    foo(x);               // calls foo(int)
    foo(uniform<int>(x)); // calls foo(uniform<int>)

    bar(x);               // calls bar(int)
    bar(uniform<int>(x)); // calls bar(int) -- uniform<int> converts implicitly to int

    baz(x);               // compiler error -- int does not convert implicitly to uniform<int>
    baz(uniform<int>(x)); // calls baz(int)
}

Reviewer: I meant the code inside the function. E.g.
void foo(int x) {
  int z = uniform<int>(x);
}
User can think that foo is called always with uniform parameter and assert that.

Author: This extension doesn't introduce that problem -- it attempts to solve it!

Today, if a parameter to a SYCL function must be uniform, that must be expressed in the function's documentation.  There is no way to describe via a function's signature that a given parameter must be uniform.  So, it's already possible to write functions that rely on assumed uniformity, like this:
```c++
void foo(sub_group sg, int x) {
  if (x == 5) {
    int z = reduce_over_group(sg, x);
  } else {
    ...
  }
}
```
...if `x` is not uniform, then it is illegal to call `reduce_over_group` inside of the conditional.  The user can document that `foo` requires `x` to be uniform, but cannot enforce it.  If they want to support both uniform- and non-uniform data, the only way to do so is via a run-time check:
```c++
void foo(sub_group sg, int x) {
  if (all_of_group(sg, x == 5)) {
    int z = reduce_over_group(sg, x);
  } else {
    ...
  }
}
```

This extension allows for the uniformity requirement to be declared as part of the function signature.  I don't think people would write the code from your example; they'd write the below instead:
```c++
void foo(uniform<int> x) {
  int z = x;
}
```

Reviewer: Creating a spec should suppose some implementation. I would like to know how you see the implementation for IA.

Reviewer: And one more question. How is this aligned with our function pointers specification? As far as I remember, we had the same intent,  explicit uniform/linear/varying specification for formal and actual parameters. Do you remember why that was not accepted? Are we going to change that spec too?

`sycl::ext::oneapi::uniform`.  Any such pointer value must point to memory
that is accessible by all work-items in the sub-group (i.e. the pointer must
point to an allocation in local or global memory).  The address space for such
pointers can be local, global or generic; if a generic pointer that points
to an allocation in private memory is passed as an argument, the behavior is
undefined.

In order to invoke the SIMD function, `invoke_simd` converts each argument
in the `args` parameter pack according to the following rules:

- Arguments of type `bool` are converted to type
  `sycl::ext::oneapi::simd_mask<bool, N>`, where `N` is the sub-group size of
  the calling kernel.  Element `i` of the SIMD type represents the value from
  the work-item with sub-group local ID `i`.

- All other arithmetic arguments of type `T` are converted to type
  `sycl::ext::oneapi::simd<T, N>`, where `N` is the sub-group size of the
  calling kernel.  Element `i` of the SIMD type represents the value from the
  work-item with sub-group local ID `i`.

- Arguments of type `sycl::ext::oneapi::uniform<T>` are converted to type `T`.

>   The value may be taken from any work-item in the sub-group; wrapping an

Reviewer: This needs to be clarified for cases when the call is under divergent condition. Calling the function under divergent condition means that not all work items are valid.

Author: Good point.

I think we should update this to say that a `uniform` can only be constructed in converged control flow (at least for now).  All of the use-cases we've seen for this tie into sub-group functions that already have this requirement.

Author: Fixed in d47e0094de20c26dfb797401a9819cb8a9dfa83b.

  argument in a `sycl::ext::oneapi::uniform` is an assertion that the variable
  holds the same value for all work-items in the sub-group. If this assertion
  does not hold, the value of the scalar variable passed to the SIMD function

>   is undefined.

Reviewer: Is there any way to indicate that for user?

Author: I'm not sure what you mean.  The user can already insert a run-time check for uniformity (using the `all_of` function) prior to taking a branch that asserts uniformity if they are concerned about correctness.

Compilers are encouraged to issue a diagnostic if a variable can _never_ be uniform, but they can't reason about the general case.  In the general case (e.g. a value read from memory), only the developer knows if a given variable is expected to hold the same value across all work-items.

- Arguments of type `std::tuple` are mapped as if each member of the tuple
  was mapped individually (e.g. a `std::tuple<float, int>` becomes a
  `std::tuple<sycl::ext::oneapi::simd<float, N>, sycl::ext::oneapi::simd<int, N>>`.

After this mapping has taken place, an appropriate variant of `Callable` is
selected based on standard overload resolution.

The return value of an `invoke_simd` function is subject to the same mapping

> rules in reverse:

Reviewer: I think this needs clarification for calls under condition. Are all elements considered as defined? Or only unmasked ones?

Author: The specification already says that `invoke_simd` can only be called in converged control flow.  Doesn't that cover this case?

Reviewer: It seems I've missed that. Thanks.

an 8-wide SIMD type `x` by a scalar value `n`.

[source, c++]
----
sycl::ext::oneapi::experimental::simd<float, 8> scale(sycl::ext::oneapi::experimental::simd<float, 8> x, float n)
{
  return x * n;
}
----

=== Invoking a SIMD Function

The `invoke_simd` free function invokes a SIMD function using all work-items
in a `sub_group`.  The `invoke_simd` interface marshals data between the
SPMD context of the calling kernel and the SIMD context of the callee,
converting arguments and return values between scalar and SIMD types as
appropriate.

[source, c++]
----
namespace sycl {
namespace ext {
namespace oneapi {
namespace experimental {

template <typename Callable, typename...T>
auto invoke_simd(sub_group sg, Callable&& f, T... args);

} // namespace experimental
} // namespace oneapi
} // namespace ext
} // namespace sycl
----

The `Callable` passed to `invoke_simd` must be a C++ callable that can be
invoked with the same number of arguments specified in the `args` parameter
pack. `Callable` may be a function object, a lambda,
or a function pointer (if the device supports `SPV_INTEL_function_pointers`).
`Callable` must be an immutable callable with the same type and state for all

> work-items in the sub-group, otherwise behavior is undefined.

Reviewer 2: A late nit: I feel like some more explanation would be helpful here: if a callable is a lambda, it is so easy to unintendedly capture something non-uniform, which will then be used inside simd function. Suggest to add non-normative comment that this particular violation of  the `immutable callable with the same type and state for all work-items in the sub-group` principle should be watched out for.

Author: Thanks, Konst.  I've added a note in ba8ff0f711becf664b61c6dada597106da44e945.

Reviewer 2: Thanks John!

implementations are expected to use the most efficient representation
of a `simd<T, N>` type (e.g. using multiple registers rather than an array).
We expect implementations of this ABI to build on Clang's `ext_vector_type`,
rather than specifying the register types directly.

The value of `N` for each `simd` argument to a SIMD function must be the same,
and represents the sub-group size of the calling kernel.  If a function accepts
`simd` arguments with multiple `N` values, it cannot be invoked from a SPMD
kernel and the implementation must issue a diagnostic in this case.

The example below shows a simple SIMD function that scales all elements of
an 8-wide SIMD type `x` by a scalar value `n`.

[source, c++]
----
sycl::ext::oneapi::experimental::simd<float, 8> scale(sycl::ext::oneapi::experimental::simd<float, 8> x, float n)
{
  return x * n;
}
----

=== Invoking a SIMD Function

The `invoke_simd` free function invokes a SIMD function using all work-items
in a `sub_group`.  The `invoke_simd` interface marshals data between the
SPMD context of the calling kernel and the SIMD context of the callee,
converting arguments and return values between scalar and SIMD types as
appropriate.

> 

Reviewer 2: Per offline discussion with @rolandschulz, the spec might need to require that the caller of `invoke_simd` should have explicitly set subgroup size - i.e. bear the `[[sycl::reqd_sub_group_size(N)]]` attribute. 

For given SPMD caller `foo` and SIMD callee `bar` the SG size could be deduced from vector arguments of the `bar` (unless `bar` signature contains uniform types only, in which case SG size does not matter), but that would present problems for SPMD callers of the `foo`, which must know if it has fixed SG size, esp. if the caller does not see `foo` body.

[source, c++]
----
namespace sycl {
namespace ext {
namespace oneapi {
namespace experimental {

template <typename Callable, typename...T>
auto invoke_simd(sub_group sg, Callable&& f, T... args);

} // namespace experimental
} // namespace oneapi
} // namespace ext
} // namespace sycl
----

The `Callable` passed to `invoke_simd` must be a C++ callable that can be
invoked with the same number of arguments specified in the `args` parameter
pack. `Callable` may be a function object, a lambda,
or a function pointer (if the device supports `SPV_INTEL_function_pointers`).
`Callable` must be an immutable callable with the same type and state for all
work-items in the sub-group, otherwise behavior is undefined.

NOTE: When using a lambda `Callable`, it is very easy for a user to
unintentionally capture non-uniform variables and violate the requirements
from the previous paragraph. This form of `invoke_simd` should be used with
caution.

> Each argument to the invoked `Callable` must be a scalar type, a SIMD type,

Reviewer 2: Re-reading this again I discovered I might be missing something. Is this really what is meant?
```suggestion
Each formal parameter of the invoked `Callable` must be a scalar type, a SIMD type,
```
otherwise (if this is really about actual arguments passed to `invoke_simd`), it is not clear what types the corresponding formal parameter will (may) have, and which part of argument conversion rules apply in this case.

Author: I didn't appreciate the difference between "argument" and "parameter" until now.  You're right, this paragraph is talking about the types that appear in the function signature of the `Callable`.  I'll make a pass through the specification and make sure we're using these words correctly.

Reviewer 2: thanks, good to know formal parameters are meant

Author: I couldn't find any other places where we'd mixed them up, but I went ahead and corrected this one in https://github.com/intel/llvm/pull/3445/commits/a00319d39431b3ade7fea34a03931a4798a72f06.

or a `std::tuple` of these types. The length of each SIMD type must match the
sub-group size of the kernel; calling a SIMD function with mismatched lengths
is an error, and an implementation must issue a diagnostic.

To enable overloading in cases where the `Callable` accepts either a scalar
type or a SIMD type, we leverage the `sycl::ext::oneapi::experimental::uniform` wrapper
type from the `SYCL_EXT_ONEAPI_UNIFORM` extension.

> 

Reviewer 2: one more late nit: it would be good to have an example how overloaded callables should be passed to invoke_simd. E.g. this won't compile - scale1 is overloaded, so `scale1` as a function pointer is ambiguous (thanks to @skozhukh for bringing up). I guess there are general C++ tools for that, but since overloading is mentioned, it would help to see how it is supposed to be done with invoke_simd.

```cpp
float scale1(sycl::ext::oneapi::simd<float, 8> x, float n) {
  return x[2] * n;
}

float scale1(sycl::ext::oneapi::simd<float, 8> x, sycl::ext::oneapi::simd<float, 8> n) {
  return x[2] * n[2];
}

...
  float x = 1;
  float n = 2;
  // invoke SIMD function
  // x values from each work-item are combined into a simd<float, 8>
  auto y = sycl::ext::oneapi::invoke_simd(
      sg, scale1, x, sycl::ext::oneapi::uniform{n});

```

Reviewer 2: ... like this:
```cpp
  auto y = sycl::ext::oneapi::invoke_simd<float(sycl::ext::oneapi::simd<float, 8>, float)>(
      sg, scale1, x, sycl::ext::oneapi::uniform{n});
```
?

Author: That would be one way to write it.  Anything that causes `scale1` to become a pointer to a specific function will work (see https://en.cppreference.com/w/cpp/language/overloaded_address).

So something like this would work, too:
```c++
auto f = static_cast<float(sycl::ext::oneapi::simd<float, 8>, float)>(scale1);
auto y = sycl::ext::oneapi::invoke_simd(sg, f, x, sycl::ext::oneapi::uniform{n});
```

There's also a neat trick where you can "lift" a function via a lambda, so that it would automatically call the correct version of `scale1` based on the arguments.  There's a good summary of that approach here: https://www.bfilipek.com/2019/03/lambdas-story-part2.html#bonus-lifting-with-lambdas.

This does make me think that the following line from the specification may be wrong:

> After this mapping has taken place, an appropriate variant of `Callable` is selected based on standard overload resolution.

@rolandschulz, what do you think about this?  It certainly looks to me like if the user passes the name of an overloaded function to `invoke_simd`, we have to resolve it _before_ we can call `invoke_simd`.  Is this a bug, or is there a trick I'm missing?

Author: I spoke to Roland offline about this.  The specification is correct, but we might want to consider adding a non-normative note and some examples to help people understand how this is expected to work.

The comment about overloading refers to situations where the `Callable` itself represents an overload set (e.g. because it is a functor with multiple call operators).  Your example doesn't fit this case, because what you're passing to `invoke_simd` is the name of an overloaded function (which decays to a function pointer).  It's the user's responsibility to handle that ambiguity.  Note that none of this is really specific to `invoke_simd`; it matches the behavior of functions like `std::invoke` and `std::apply`.

There's no right answer here, but we should adopt a consistent style for training examples.  If the user doesn't want `invoke_simd` to automatically select an overload, the clearest syntax is the one that you proposed:
```c++
  auto y = sycl::ext::oneapi::invoke_simd<float(sycl::ext::oneapi::simd<float, 8>, float)>(sg, scale1, x, sycl::ext::oneapi::uniform{n});
```

If the user does want `invoke_simd` to automatically select an overload, converting the functions into a functor with multiple call operators is the clearest syntax:
```c++
struct scale1 {
  float operator()(sycl::ext::oneapi::simd<float, 8> x, float n) {
    return x[2] * n;
  }

  float operator()(sycl::ext::oneapi::simd<float, 8> x, sycl::ext::oneapi::simd<float, 8> n) {
    return x[2] * n[2];
  }
}
...
  auto y = sycl::ext::oneapi::invoke_simd(sg, scale1(), x, sycl::ext::oneapi::uniform{n});
```
...although note that we now have to pass an instance of `scale1` to `invoke_simd` instead of just the name.

Reviewer 2: Thanks for the explanation, John.

> but we might want to consider adding a non-normative note and some examples to help people understand how this is expected to work

yes, please

Author: Added in https://github.com/intel/llvm/pull/3445/commits/b278fb495b8ce763667dc8cacfe1ebbc1c2b8d08.

Each argument in the `args` parameter pack must be an arithmetic type, a
trivially copyable type wrapped in a `sycl::ext::oneapi::experimental::uniform`, or a
`std::tuple` of these types.  Arguments may not be pointers or references,
but pointers (like any other non-arithmetic type) may be passed if wrapped in a

> `sycl::ext::oneapi::experimental::uniform`.  Any such pointer value must point to memory

Reviewer 2: This basically forbids to have a simd callee with vector-of-pointers argument:
```cpp
void simd_foo(simd<int*,16> ptrs, ...) { scatter(ptrs,...); }
```
And the spec implies non-uniform pointer is represented as a base pointer + a vector of offsets, so SIMD library functions must be written with this style to be callable from SYCL:
```cpp
void simd_foo(int *base, simd<uintptr_t,16> offsets, ...) { ... }
```
This is natural for many platforms, but this implication should better be explicitly mentioned, I think.

(this and my other late comments can be addressed in future revisions of the spec, not a blocker for merge)

Author: I don't think this is a significant usability issue, because you can write a wrapper:
```c++
void simd_foo_from_sycl(int *base, simd<uintptr_t,16> offsets, ...) {
  simd<int*, 16> ptrs(base + offsets);
  simd_foo(ptrs, ...);
}
```

Would you prefer a non-normative note pointing out this is possible, or an unresolved issue suggesting we should revisit this restriction?

Reviewer 2: Ah, you are right, thanks. As simple as that :)

> Would you prefer a non-normative note pointing out this is possible, or an unresolved issue suggesting we should revisit this restriction?

Non-normative note would be enough.

Author: Added in https://github.com/intel/llvm/pull/3445/commits/b91567f38984c49f529b0ab2dc215fce83563f89.

We expect implementations of this ABI to build on Clang's `ext_vector_type`,
rather than specifying the register types directly.

The value of `N` for each `simd` argument to a SIMD function must be the same,
and represents the sub-group size of the calling kernel.  If a function accepts
`simd` arguments with multiple `N` values, it cannot be invoked from a SPMD
kernel and the implementation must issue a diagnostic in this case.

The example below shows a simple SIMD function that scales all elements of
an 8-wide SIMD type `x` by a scalar value `n`.

[source, c++]
----
sycl::ext::oneapi::simd<float, 8> scale(sycl::ext::oneapi::simd<float, 8> x, float n)
{
  return x * n;
}
----

=== Invoking a SIMD Function

The `invoke_simd` free function invokes a SIMD function using all work-items
in a `sub_group`.  The `invoke_simd` interface marshals data between the
SPMD context of the calling kernel and the SIMD context of the callee,
converting arguments and return values between scalar and SIMD types as
appropriate.

[source, c++]
----
namespace sycl {
namespace ext {
namespace oneapi {

template <typename Callable, typename...T>

> auto invoke_simd(sub_group sg, Callable&& f, T... args);

Reviewer 2: (repeating the Q from internal discussion here)

Did you have some particular implementation approach in mind?

I’m thinking that invoke_simd will be an intrinsic which will capture the callable, arguments and argument/ret uniformity information (not 100% sure if it is needed). So that SPMD BE could do proper transformation and replace the call to invoke_simd with actual call after widening the caller.

If you had something similar in mind, then Ideally this would be a SPIRV intrinsic – is there a plan to standardize this part?

Author: I suspect that our implementation will look very similar to what you've described, at least for the backends that consume SPIR-V.  We need _something_ at the SPIR-V level to convey that we're transitioning from a SPMD context to a SIMD context, and a special `OpFunctionCall` seems like a good way to do that.  If we introduce such a thing, I think we should standardize it.

I don't think this specification should depend on it, though.  There are other valid implementations, and backends that don't go through SPIR-V should have the freedom to do something else.

Reviewer 2: Good, thanks.

> I don't think this specification should depend on it, though.

Sure, agreed.

Since this execution model guarantees SIMD-like behavior, there is no need for
the user to insert any form of explicit synchronization functions to ensure
memory consistency across SIMD lanes.

== Issues

. Should we allow reference arguments?
+
--
*RESOLVED*: No.  Unlike pointers, it is less clear that references carry
address space information, and users may not expect a +T&+ argument to behave
like a pointer to the generic address space.  Users are also more likely to try
and pass a reference to a variable in the private address space than a reference
to a variable in the local or global address spaces, which would lead to
undefined behavior and issues that are difficult to debug.
--

. How should returning multiple values and passing user-defined structs be handled?
+
--
*UNRESOLVED*: +std::tuple+ addresses both issues from a user interface perspective,
but this aspect of the design may need to be revisited as we gain implementation experience.
--

. Should arguments require trivially copyable or device copyable?
+
--
*UNRESOLVED*: Device copyable permits bitwise copies of a wider range of classes, but the

> SYCL 2020 specification states that this only applies to inter-device transfers.

Author: I'd be interested to hear from @rolandschulz and @nliber on this.

The paragraphs from SYCL 2020 that this extension is referencing are these two (from https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec::device.copyable):
> The SYCL implementation may need to copy data between the host and a device or between two devices
> ...
> When doing an inter-device transfer of an object of type `T`, the effect of each eligible copy constructor, move constructor, copy assignment operator, and move assignment operator is the same as a bitwise copy of the object;

When the specification says "inter-device" here, I think it's classifying the host as a device.  I agree that something like "a transfer between host and device or between two devices" would by clearer.

The reason that I don't believe we can use device copyable for this extension remains unchanged, though; my reading of the paragraph above is that the constructors and operators only behave as bitwise copies when moving an object across a host-device or device-device boundary.  An implementation may still execute constructors and copy operators in the device code.

> SYCL 2020 specification states that this only applies to inter-device transfers.

Reviewer 3: > The reason that I don't believe we can use device copyable for this extension remains unchanged, though; my reading of the paragraph above is that the constructors and operators only behave as bitwise copies when moving an object across a host-device or device-device boundary.  An implementation may still execute constructors and copy operators in the device code.

Is it possible for user to write two different constructors: one which is invoked during host-device communication and another one which is invoked only on device side? I guess it should be possible with `__SYCL_DEVICE_ONLY__`. Do we really want to support that? Is there any particular use case where two versions of constructors are needed?

> SYCL 2020 specification states that this only applies to inter-device transfers.

Author: > Is it possible for user to write two different constructors: one which is invoked during host-device communication and another one which is invoked only on device side? I guess it should be possible with __SYCL_DEVICE_ONLY__. Do we really want to support that? Is there any particular use case where two versions of constructors are needed?

I think that would be an ODR violation, and so wouldn't be allowed.

The "When doing an inter-device transfer of an object of type `T`" is what's confusing me, here, but you're right -- if there can only be one definition of each of those constructors and operators, it's unclear how a user could specialize the functions for that use-case.  But I'm going to stop guessing, and wait for the experts (@nliber and @rolandschulz) to weigh in.

> SYCL 2020 specification states that this only applies to inter-device transfers.

Reviewer 4: > I think that would be an ODR violation, and so wouldn't be allowed.

ODR violation apply to a single compiler. We aren't saying that this isn't allowed for split compilation (but we also not clearly saying it is allowed).

> The reason that I don't believe we can use device copyable for this extension remains unchanged, though; my reading of the paragraph above is that the constructors and operators only behave as bitwise copies when moving an object across a host-device or device-device boundary. An implementation may still execute constructors and copy operators in the device code.

I agree this is a correct reading of the wording. I don't think we spent any time wondering whether it should apply to more than device transfer because that was the use-case this was added for.

I'm not aware of any use case of device-copyable for which the copy constructor/assignment on the device does anything different from bitwise copy. And if we changed the requirement for the user to declare a type device-copyable that it doesn't do anything extra on the device we could use it here. But I think it would require a change of that paragraph to clarify it.

> SYCL 2020 specification states that this only applies to inter-device transfers.

Author: Thanks, Roland.

@AlexeySachkov: Given Roland's answer, I propose that we merge this extension specification with the question about `is_device_copyable` unresolved.  I'll open a Khronos issue to discuss the intent of the specification and the need for potential clarification, and we can revisit this later.  Does that work for you?

> SYCL 2020 specification states that this only applies to inter-device transfers.

Reviewer 3: > @AlexeySachkov: Given Roland's answer, I propose that we merge this extension specification with the question about is_device_copyable unresolved. I'll open a Khronos issue to discuss the intent of the specification and the need for potential clarification, and we can revisit this later. Does that work for you?

Sure, no objections

in a `sub_group`.  The `invoke_simd` interface marshals data between the
SPMD context of the calling kernel and the SIMD context of the callee,
converting arguments and return values between scalar and SIMD types as
appropriate.

[source, c++]
----
namespace sycl {
namespace ext {
namespace oneapi {
namespace experimental {

template <typename Callable, typename...T>
auto invoke_simd(sub_group sg, Callable&& f, T... args);

} // namespace experimental
} // namespace oneapi
} // namespace ext
} // namespace sycl
----

The `Callable` passed to `invoke_simd` must be a C++ callable that can be
invoked with the same number of arguments specified in the `args` parameter
pack. `Callable` may be a function object, a lambda,
or a function pointer (if the device supports `SPV_INTEL_function_pointers`).
`Callable` must be an immutable callable with the same type and state for all
work-items in the sub-group, otherwise behavior is undefined.

NOTE: When using a lambda `Callable`, it is very easy for a user to
unintentionally capture non-uniform variables and violate the requirements
from the previous paragraph. This form of `invoke_simd` should be used with
caution.

> Each parameter of the invoked `Callable` must be a scalar type, a SIMD type,

Reviewer 2: ```suggestion
Each formal parameter of the invoked `Callable` must be a scalar type, a SIMD type,
```
